src/aioswitcher/api/__init__.py

```diff
@@ -103,12 +108,21 @@ def __init__(
         self._device_key = device_key
         self._port = port
         self._connected = False
+        if token:
+            self._token = tok.he(token)  # type: ignore
+        else:
+            self._token = ""
 
     @property
     def connected(self) -> bool:
         """Return true if api is connected."""
         return self._connected
 
+    @property
+    def is_using_token(self) -> bool:
+        """Return true if token is used."""
+        return bool(self._device_type.token_needed) and str(self._token) != ""
+
```

Comment:
I don't understand the logic here, if device require token we need to have a token and if it is missing the library should raise an errror

Reply:
Will try that and update..

Reply:
I changed the logic. and add tests for the token.
Please review again

Reply:
The refactoring didn't change anything in the logic. I will ask again. If token is needed and not supplied in the API call you are just skipping the token.
Please also don't put the functions related to the token in the device, they should be in this class (API).
I think you shouldn't validate at all if token is needed as this can be responsibility of the caller, however if you do, you should raise if token is needed and not supplied, not just skip it.

Reply:
I can't get back to my previous state.. too many changes. I will figure out this issue alone and fix it.
After that issue is fixed I will be in a code freeze until the PR is fully reviewed.